### PR TITLE
Refactor chat template and support accurate name matching.

### PR DESCRIPTION
**Pull request by AllentDan**
- [x] Refactor chat template. Removed `from_config`, `decorate_prompt`, `_translate_messages`, and `sampling_param` from the chat template.
- [x] accurate name matching

### CODE REVIEW

**Comment by AllentDan** [comment]
Need update after #1168 merged

**Comment by lvhan028** [comment]
#1168 has been merged

File: requirements/runtime.txt [hunk 1]
```diff
@@ -1,6 +1,5 @@
 fastapi
 fire
-fuzzywuzzy
```
Comment (lvhan028) [inline requirements/runtime.txt:3]: may also remove `fuzzywuzzy` in `readthedocs.txt`

File: lmdeploy/model.py [hunk 2]
```diff
@@ -43,6 +36,7 @@ class ChatTemplateConfig:
     eoh: Optional[str] = None
     assistant: Optional[str] = None
     eoa: Optional[str] = None
+    stop_word_suffix: Optional[str] = None
```
Comment (lvhan028) [inline lmdeploy/model.py:39]: why `suffix`?

**Comment by lvhan028** [comment]
比较想把这个修改加入到 v0.2.5

File: lmdeploy/model.py (hunk 2 above)
Comment (AllentDan) [reply lmdeploy/model.py:39]: For interactive prompts. For example, `eoa='<eoa>'`, `stop_word_suffix='\n'`

Comment (lvhan028) [reply lmdeploy/model.py:39]: The name is confusing. Are you suggesting a separator after `eoa`?

Comment (AllentDan) [reply lmdeploy/model.py:39]: Yeah, the previous name is `eoa_suffix`. However, I found the following case:
```python
                 eoa='<|im_end|>',
                 stop_word_suffix='\n',
                 stop_words=['<|im_end|>', '<|action_end|>'],
```
Shall I rename it back?

Comment (lvhan028) [reply lmdeploy/model.py:39]: How about a `separator`?

File: examples/vl/qwen_model.py [hunk 3]
```diff
@@ -45,27 +37,30 @@ def _concat_image_info(self, prompt):
...
+                       system=self.eosys)
+        ret = ''
+        for message in messages:
+            role = message['role']
+            content = message['content']
+            if role == 'user' and not isinstance(content):
+                content = [content[0]['text'], len(content) - 1]
+                content = self._concat_image_info(content)
+            ret += f'{box_map[role]}{content}{eox_map[role]}'
+        ret += f'{self.assistant}'
```
Comment (irexyc) [inline examples/vl/qwen_model.py:63]: messages2prompt 是说messages里面没有system的话，就不加了是么？这个之前貌似默认是有的。

File: lmdeploy/model.py [hunk 4]
```diff
@@ -212,20 +166,61 @@ def messages2prompt(self, messages, sequence_start=True):
...
+    """Chat template of vicuna model."""
+
+    def __init__(
+            self,
+            meta_instruction="""A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions.\n""",  # noqa: E501
+            user='USER: ',
+            eoh='\n',
+            assistant='ASSISTANT: ',
+            eoa='</s>',
+            separator='\n',
```
Comment (irexyc) [inline lmdeploy/model.py:196]: vicuna 的模版里面有 `\n`  么？

File: examples/vl/qwen_model.py (hunk 3 above)
Comment (irexyc) [reply examples/vl/qwen_model.py:63]: 好像也不是都有，baichuan2-7b 之前也没管system

File: lmdeploy/model.py [hunk 5]
```diff
@@ -837,137 +708,80 @@ class UltraChat(BaseModel):
...
-            content = message['content']
-            ret += f'{eval(f"self.{role}")}{content}{eox_map[role]}'
-        ret += f'{self.assistant}'
-        return ret
-
-
-@MODELS.register_module(name=['Mistral-7B-Instruct', 'Mixtral-8x7B-Instruct'])
+@MODELS.register_module(name=['mistral-instruct', 'mixtral-instruct'])
 class MistralChat(BaseModel):
     """Template of Mistral and Mixtral Instruct models.
```
Comment (irexyc) [inline lmdeploy/model.py:786]: `class MistralChat(BaseChatTemplate):` ？

Comment (irexyc) [reply lmdeploy/model.py:786]: Deepseek， gemma

File: lmdeploy/model.py (hunk 4 above)
Comment (irexyc) [reply lmdeploy/model.py:196]: vicuna decorate之后跟之前的不一样，可以看一下

File: examples/vl/qwen_model.py [hunk 6]
```diff
@@ -45,27 +37,30 @@ def _concat_image_info(self, prompt):
...
+        eox_map = dict(user=self.eoh,
+                       assistant=self.eoa + self.stop_word_suffix,
+                       system=self.eosys)
+        ret = ''
+        for message in messages:
+            role = message['role']
+            content = message['content']
+            if role == 'user' and not isinstance(content):
+                content = [content[0]['text'], len(content) - 1]
+                content = self._concat_image_info(content)
```
Comment (irexyc) [inline examples/vl/qwen_model.py:61]: if role == 'user' and not isinstance(content, str):

这个之前应该写错了，不过没有用到。

File: lmdeploy/model.py [hunk 7]
```diff
@@ -212,20 +166,61 @@ def messages2prompt(self, messages, sequence_start=True):
...
+                       system=self.system)
+        eox_map = dict(user=self.eoh,
+                       assistant=self.eoa + self.separator,
+                       system=self.eosys)
+        ret = ''
+        for message in messages:
+            role = message['role']
+            content = message['content']
+            ret += f'{box_map[role]}{content}{eox_map[role]}'
+        ret += f'{self.assistant}'
```
Comment (irexyc) [inline lmdeploy/model.py:180]: messages2prompt 是说messages里面没有system的话，就不加了是么？这个之前貌似默认是有的。

Comment (AllentDan) [reply lmdeploy/model.py:180]: 对，这个行为被BC了（internlm 的模型除外），需要讨论下。之前如果用户不传 meta_instrucion，会用默认的。现在只看用户传的，如果用户没传就没在用。

File: examples/vl/qwen_model.py [hunk 8]
```diff
@@ -45,27 +37,30 @@ def _concat_image_info(self, prompt):
...
+                       assistant=self.assistant,
+                       system=self.system)
+        eox_map = dict(user=self.eoh,
+                       assistant=self.eoa + self.stop_word_suffix,
+                       system=self.eosys)
+        ret = ''
+        for message in messages:
+            role = message['role']
+            content = message['content']
+            if role == 'user' and not isinstance(content):
```
Comment (grimoire) [inline examples/vl/qwen_model.py:59]: isinstance require two inputs

File: lmdeploy/model.py (hunk 4 above)
Comment (AllentDan) [reply lmdeploy/model.py:196]: 是有点奇怪，FastChat 文档里面的模板和代码效果略有差异。[文档](https://github.com/lm-sys/FastChat/blob/main/docs/vicuna_weights_version.md#example-prompt-weights-v11-v13-v15)里效果是这样的：
```
A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions.

USER: Hello!
ASSISTANT: Hello!</s>
USER: How are you?
ASSISTANT: I am good.</s>
```
然后代码效果这样的：
```
A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions. USER: Hello! ASSISTANT: Hi!</s>USER: How are you? ASSISTANT:
```

**Comment by grimoire** [comment]
`tokenizer_config.json` contain a lot of useful information might help to ease the chat template.
For example [gemma-7b-it](https://huggingface.co/google/gemma-7b-it) has `bos_token`, `eos_token` and `chat_template`.

File: lmdeploy/cli/cli.py [hunk 9]
```diff
@@ -103,16 +103,33 @@ def convert(args):
     def list(args):
         """List the supported model names."""
         engine = args.engine
-        assert engine in ['turbomind', 'pytorch']
+        assert engine in ['turbomind', 'pytorch', None]
```
Comment (lvhan028) [inline lmdeploy/cli/cli.py:106]: Let's list all the supported models without telling engines apart

**Comment by lvhan028** [comment]
May merge the latest main to resolve conflicts and readthedocs issue

File: lmdeploy/cli/utils.py [hunk 10]
```diff
@@ -261,7 +261,7 @@ def engine(parser):
 
         return parser.add_argument('--engine',
                                    type=str,
-                                   default='turbomind',
+                                   default=None,
```
Comment (RunningLeon) [inline lmdeploy/cli/utils.py:264]: this change will affect the behavers of all cli. You should add an argument of `default` in line 259 and pass `default=None` when calling it in `add_parser_list`

**Review by grimoire — APPROVED** [review]

File: lmdeploy/model.py [hunk 11]
```diff
@@ -53,7 +47,13 @@ def chat_template(self):
             for key, value in dataclasses.asdict(self).items()
             if value is not None
         }
-        model: BaseModel = MODELS.get(self.model_name).from_config(**attrs)
+        if self.model_name in MODELS.module_dict.keys():
+            model: BaseModel = MODELS.get(self.model_name)(**attrs)
+        else:
+            logger.warning(
+                f'Could not find {self.model_name} in registered models. '
```
Comment (lvhan028) [inline lmdeploy/model.py:54]: f"unable to locate '{self.model_name}' in registered chat templates, falling back to the default template. Please register a new chat template if the default one does not meet your requirements”

Comment (AllentDan) [reply lmdeploy/model.py:54]: This is for json template usage. Any new name will be treated as newly registered.

File: lmdeploy/model.py [hunk 12]
```diff
@@ -194,12 +145,15 @@ def decorate_prompt(self, prompt, sequence_start=True):
         Returns:
             str: the concatenated prompt
         """
-        assert self.capability == 'chat', \
-            f'{type(self).__name__} has no capability of {self.capability}'
+        if self.capability == 'completion':
+            return prompt
         if sequence_start:
-            return f'{self.system} {self.user}: {prompt} {self.assistant}: '
+            return f'{self.system}{self.meta_instruction}{self.eosys}' \
```
Comment (lvhan028) [inline lmdeploy/model.py:151]: Should we consider "self.separator"?

File: lmdeploy/model.py (hunk 7 above)
Comment (lvhan028) [reply lmdeploy/model.py:180]: 确实有些纠结。这里修改的出发点是，可以简化些逻辑，不用translate_messages，对吧。

Comment (AllentDan) [reply lmdeploy/model.py:180]: > 确实有些纠结。这里修改的出发点是，可以简化些逻辑，不用translate_messages，对吧。

对，其实我们是不知道ChatGPT内部的行为是什么。不过像 FastChat 似乎内部是用了缺省值的，如果用户没传的话。

File: lmdeploy/model.py [hunk 13]
```diff
@@ -358,75 +355,69 @@ def __init__(self,
...
 
+    @classmethod
+    def match(cls, model_path: str) -> Optional[str]:
+        """Return the model_name that was registered to MODELS.
+
+        Args:
+            model_path (str): the model path used for matching.
+        """
+        path = model_path.lower()
+        if 'internlm2' in path and 'chat' in path:
```
Comment (lvhan028) [inline lmdeploy/model.py:370]: internlm2-math 的对话模型和 internlm2-chat 一致

File: tests/test_lmdeploy/test_model.py [hunk 14]
```diff
@@ -1,20 +1,19 @@
 import pytest
```
Comment (RunningLeon) [inline tests/test_lmdeploy/test_model.py:1]: ut failed with baichuan

File: lmdeploy/model.py [hunk 15]
```diff
@@ -440,242 +431,135 @@ def __init__(self,
...
+        if 'qwen' in model_path.lower():
+            return 'qwen-chat'
 
 
 @MODELS.register_module(name='codellama')
 class CodeLlama(Llama2):
 
     def __init__(self,
-                 system='',
+                 meta_instruction='',
```
Comment (RunningLeon) [inline lmdeploy/model.py:537]: it seems `session_len`, `suffix_first` and `stop_words` are unused ...

Comment (AllentDan) [reply lmdeploy/model.py:537]: Because `codellama` can do completion, infilling, and python tasks apart from chat.

**Review by RunningLeon — APPROVED** [review]
LGTM

**Comment by lvhan028** [comment]
I am doing BC-breaking test, comparing the results between v0.2.5 and this PR. They are different

```python
prompt = 'hi, how are you'

from lmdeploy.model import MODELS

chat_templates = [
    'internlm', 'llama', 'base',
    'wizardlm', 'vicuna',
    'internlm-chat', 'internlm-chat-7b', 'internlm-chat-20b', 'internlm-chat-7b-8k',
    'internlm2-chat', 'internlm2-chat-1_8b', 'internlm2-chat-7b', 'internlm2-chat-20b',
    'baichuan-7b',
    'baichuan2-7b',
    'llama2', 'llama-2', 'llama-2-chat',
    'qwen-7b', 'qwen-14b',
    'codellama',
    'falcon',
    'chatglm2-6b',
    'solar', 'solar-70b',
    'ultracm',
    'ultralm',
    'yi', 'yi-chat', 'yi-200k', 'yi-34b',
    'Mistral-7B-Instruct', 'Mixtral-8x7B-Instruct',
    'gemma',
    'deepseek-chat'
]

for _template in chat_templates:
    print(f'---------{_template}---------')
    model = MODELS.get(_template)()
    print(model.get_prompt(prompt))

```

File: examples/vl/qwen_model.py [hunk 16]
```diff
@@ -19,20 +19,12 @@ def __init__(self,
                  top_p=0.3,
                  top_k=None,
                  temperature=1.0,
-                 im_start='<|im_start|>',
```
Comment (lvhan028) [inline examples/vl/qwen_model.py:22]: 名字改为 qwen-vl 吧

**Comment by AllentDan** [comment]
还有个问题，base 模型的匹配，现在基本都会匹配不到名字。

File: lmdeploy/model.py [hunk 17]
```diff
@@ -837,364 +665,180 @@ class UltraChat(BaseModel):
...
-                else:
-                    ret += f'{self._eos}{ret}'
-            elif role == 'assistant':
-                ret += f' {content}\n\n'
-            else:
-                raise Exception('Only have user and '
-                                'assistant roles, '
-                                f'but given: {role}')
-        return ret
+        if 'deepseek' in model_path.lower():
```
Comment (lvhan028) [inline lmdeploy/model.py:829]: deepseek coder 和 deepseek chat 模型对话模板不一样

Comment (AllentDan) [reply lmdeploy/model.py:829]: 我们两个都支持吗？

Comment (lvhan028) [reply lmdeploy/model.py:829]: deepseek-coder后面可以加。
我想表达的是，`if 'deepseek' in model_path.lower():` 不严谨，deepseek-coder 也能match

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**